From 20e5d1725f983e86c1270da5f0dd8a1b8f005c29 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Wed, 11 Oct 2023 16:59:30 +0200
Subject: [PATCH 0299/1695] plat-stm32mp2: add platform-specific abort handler

When a data abort occurs and its fault type is FAULT_TYPE_IGNORE, it
may be an abort generated by the SERC hardware block. Check if a
SERC Illegal Access was caught and print the SERC register and panic()
if that is the case.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git d6a1c4ba0fce2a1925f32acd471eb26ace964855 plat-stm32mp2: add platform-specific abort handler]
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: I8e9ae8494c77a594ea85c9cb80e22813a7c1f80e
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/333125
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/arch/arm/plat-stm32mp2/main.c |  8 ++++++++
 core/drivers/firewall/stm32_serc.c | 18 +++++++++++++++---
 core/include/drivers/stm32_serc.h  |  7 +++++++
 3 files changed, 30 insertions(+), 3 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp2/main.c b/core/arch/arm/plat-stm32mp2/main.c
index ba048fddf..d7c2e570d 100644
--- a/core/arch/arm/plat-stm32mp2/main.c
+++ b/core/arch/arm/plat-stm32mp2/main.c
@@ -9,8 +9,10 @@
 #include <drivers/stm32_rif.h>
 #include <drivers/stm32_risab.h>
 #include <drivers/stm32_risaf.h>
+#include <drivers/stm32_serc.h>
 #include <drivers/stm32_uart.h>
 #include <initcall.h>
+#include <kernel/abort.h>
 #include <kernel/boot.h>
 #include <kernel/dt.h>
 #include <kernel/interrupt.h>
@@ -164,3 +166,9 @@ void stm32_rif_access_violation_action(void)
 #endif
 }
 #endif /* CFG_STM32_RIF */
+
+void plat_abort_handler(struct thread_abort_regs *regs __unused)
+{
+	/* If fault is ignored, it could be due to a SERC event */
+	stm32_serc_handle_ilac();
+}
diff --git a/core/drivers/firewall/stm32_serc.c b/core/drivers/firewall/stm32_serc.c
index ef902a3e9..3c4a646b7 100644
--- a/core/drivers/firewall/stm32_serc.c
+++ b/core/drivers/firewall/stm32_serc.c
@@ -110,12 +110,13 @@ static TEE_Result stm32_serc_parse_fdt(void)
 #define SERC_EXCEPT_MSB_BIT(x) (x * _PERIPH_IDS_PER_REG + _PERIPH_IDS_PER_REG - 1)
 #define SERC_EXCEPT_LSB_BIT(x) (x * _PERIPH_IDS_PER_REG)
 
-static enum itr_return stm32_serc_itr(struct itr_handler *h __unused)
+void stm32_serc_handle_ilac(void)
 {
 	struct serc_driver_data *ddata = serc_dev.ddata;
 	struct stm32_serc_platdata *pdata = &serc_dev.pdata;
 	uintptr_t base = pdata->base;
 	int nreg = DIV_ROUND_UP(ddata->num_ilac, _PERIPH_IDS_PER_REG);
+	bool do_panic = false;
 	uint32_t isr = 0;
 	uint32_t i = 0;
 
@@ -126,10 +127,13 @@ static enum itr_return stm32_serc_itr(struct itr_handler *h __unused)
 		isr = io_read32(base + _SERC_ISR0 + offset);
 		isr &= io_read32(base + _SERC_IER0 + offset);
 
-		if (isr)
+		if (isr) {
+			do_panic = true;
+
 			EMSG("SERC exceptions [%d:%d]: %#x",
 			     SERC_EXCEPT_MSB_BIT(i), SERC_EXCEPT_LSB_BIT(i),
 			     isr);
+		}
 
 		while (isr && tries < _PERIPH_IDS_PER_REG) {
 			EMSG("SERC exception ID: %d",
@@ -145,8 +149,16 @@ static enum itr_return stm32_serc_itr(struct itr_handler *h __unused)
 		}
 	}
 
-	if (IS_ENABLED(CFG_STM32_PANIC_ON_SERC_EVENT))
+	if (do_panic) {
 		stm32_rif_access_violation_action();
+		if (IS_ENABLED(CFG_STM32_PANIC_ON_SERC_EVENT))
+			panic();
+	}
+}
+
+static enum itr_return stm32_serc_itr(struct itr_handler *h __unused)
+{
+	stm32_serc_handle_ilac();
 
 	return ITRR_HANDLED;
 }
diff --git a/core/include/drivers/stm32_serc.h b/core/include/drivers/stm32_serc.h
index 72a40e2e8..0571c261f 100644
--- a/core/include/drivers/stm32_serc.h
+++ b/core/include/drivers/stm32_serc.h
@@ -18,4 +18,11 @@ struct stm32_serc_platdata {
 	int irq;
 };
 
+/* Helper to print and handle SERC ILACs */
+#ifdef CFG_STM32_SERC
+void stm32_serc_handle_ilac(void);
+#else /* CFG_STM32_SERC */
+static inline void stm32_serc_handle_ilac(void) { };
+#endif /* CFG_STM32_SERC */
+
 #endif /* __DRIVERS_STM32_SERC_H__ */
-- 
2.47.2

